consistent maintenance mode strategy behavior

m-ildefons opened the pull request:
The Pod deletion filter identifies the Pods belonging to VMs, which need to be deleted during a node drain by looking at their maintenance-mode strategy label. The default value for this label is `Migrate`, which indicates that the Pod should be deleted during the node drain. Pods with an invalid label, i.e. where the value of the maintenance-mode strategy label is not one of:
  - Migrate
  - ShutdownAndRestartAfterEnable
  - ShutdownAndRestartAfterDisable
  - Shutdown are now also being treated with the default behavior. This fixes the problem that if a VM contains an invalied value in this label, nodes can become stuck in `Cordoned` state when transitioning to maintenance mode, as the node drain controller won't shut down the VM, but it's also not migrated away from the node. Therefore the VM keeps running, preventing the node from completely transitioning into maintenance mode.

Also adds a check to the validating webhook for the VirtualMachine resources.
This check also ensures that updates to virtual machines can not update the label to an invalid value that is none of the above mentioned.
However the check still alows VirtualMachines to be created without the maintenance mode strategy label, since external integrators like Rancher don't always set this label.

For backwards compatibility this webhook allows VirtualMachine updates to a VirtualMachine if the old VirtualMachine had an invalid maintenance mode strategy label and the new VirtualMachine does not change this label.
This case is crucial to allow existing VirtualMachines to continue working as expected. The webhook will emit a suitable log message.

related-to: #6835

#### Problem:

The scope of https://github.com/harvester/harvester/pull/9528 makes review too hard to complete in time for v1.7.0

#### Solution:

This PR contains a smaller and less controversial subset of the changes of PR#9528. This should make review easier and implement at least most of the necessary functionality.

#### Related Issue(s):

#6835

---

# Test plan:

## Scenario 1

Ensure that new VirtualMachines have a valid maintenance-mode strategy or none at all.

### Test 1

No maintenance-mode strategy given.

#### Setup:

Create a new VirtualMachine and specify no maintenance-mode strategy:
```
type: kubevirt.io.virtualmachine
metadata:
  namespace: default
  labels:
    harvesterhci.io/creator: harvester
    harvesterhci.io/os: linux
  name: test
[...]
```

#### Expected Result

The creation of the VM should be accepted.

### Test 2

Valid maintenance-mode strategy given.

#### Setup:

Create a new VirtualMachine and specify a valid maintenance-mode strategy:
```
type: kubevirt.io.virtualmachine
metadata:
  namespace: default
  labels:
    harvesterhci.io/maintain-mode-strategy: Shutdown
    [...]
  name: test
[...]
```

#### Expected Result

The previously specified maintenance-mode strategy should be persisted in the VirtualMachine resource's metadata:
```
type: kubevirt.io.virtualmachine
metadata:
  namespace: default
  labels:
    harvesterhci.io/maintain-mode-strategy: Shutdown
    [...]
  name: test
[...]
```

## Scenario 2

If an invalid maintenance-mode strategy is specified, the VirtualMachine resource should be rejected.

### Test 1

Invalid maintenance-mode strategy given.

#### Setup:

Create a new VirtualMachine and specify an invalid maintenance-mode strategy:
```
type: kubevirt.io.virtualmachine
metadata:
  namespace: default
  labels:
    harvesterhci.io/maintain-mode-strategy: foobar
    [...]
  name: test
[...]
```

#### Expected Result

The creation should be rejected with a sensible error message

```
The request is invalid: metadata.labels[harvesterhci.io/maintain-mode-strategy]: invalid maintenance mode strategy: foobar
```

## Scenario 3

Updates to a VirtualMachine should be allowed if the new maintenance mode strategy is valid.

### Test 1

Valid maintenance-mode strategy given.
```
type: kubevirt.io.virtualmachine
metadata:
  namespace: default
  labels:
    harvesterhci.io/maintain-mode-strategy: Migrate
    [...]
  name: test
[...]
```

#### Expected Result

The maintenance mode strategy should be successfully updated to the new value

### Test 2

Invalid maintenance-mode strategy given.
```
type: kubevirt.io.virtualmachine
metadata:
  namespace: default
  labels:
    harvesterhci.io/maintain-mode-strategy: foobar
    [...]
  name: test
[...]
```

#### Expected Result

The update to the VirtualMachine should be rejected on the basis that the new maintenance mode strategy is invalid

## Scenario 4

Ensure that existing VMs don't break.

Setup:
- Install Harvester without this change
- Create one VirtualMachine (`testvm-no-label`) without a maintenance mode strategy label
- Create one VirtualMachine (`testvm-invalid-label`) with an invalid maintenance mode strategy label
- Update to Harvester with this change

### Test 1

Make an update to both VMs (`testvm-no-label`, `testvm-invalid-label`) that does not modify the maintenance mode strategy label

#### Expected Result

The updates to the VirtualMachines should be successfully accepted and applied

### Test 2

Update the maintenance mode strategy label on both VMs (`testvm-no-label`, `testvm-invalid-label`)  to a new invalid value.

#### Expected Result

The updates to the VirtualMachines should be rejected because the new maintenance mode strategy is invalid

### Test 3

Update the maintenance mode strategy label on both VMs (`testvm-no-label`, `testvm-invalid-label`)  to a new valid value.

#### Expected Result

The updates to the VirtualMachines should be successfully accepted and applied

#### Additional documentation or context

m-ildefons committed:
consistent maintenance mode strategy behavior

The Pod deletion filter identifies the Pods belonging to VMs, which need
to be deleted during a node drain by looking at their maintenance-mode
strategy label. The default value for this label is `Migrate`, which
indicates that the Pod should be deleted during the node drain.
Pods with an invalid label, i.e. where the value of the maintenance-mode
strategy label is not one of:
  - Migrate
  - ShutdownAndRestartAfterEnable
  - ShutdownAndRestartAfterDisable
  - Shutdown
are now also being treated with the default behavior.
This fixes the problem that if a VM contains an invalied value in this
label, nodes can become stuck in `Cordoned` state when transitioning to
maintenance mode, as the node drain controller won't shut down the VM,
but it's also not migrated away from the node. Therefore the VM keeps
running, preventing the node from completely transitioning into
maintenance mode.

Also adds a check to the validating webhook for the VirtualMachine
resources.
This check also ensures that updates to virtual machines can not update
the label to an invalid value that is none of the above mentioned.
However the check still alows VirtualMachines to be created without the
maintenance mode strategy label, since external integrators like Rancher
don't always set this label. And it also allows the removal of the
label, which is functionally equivalent to setting it to Migrate.

For backwards compatibility this webhook allows VirtualMachine updates
to a VirtualMachine if the old VirtualMachine had an invalid maintenance
mode strategy label and the new VirtualMachine does not change this
label.
This case is crucial to allow existing VirtualMachines to continue
working as expected. The webhook will emit a suitable log message.

related-to: #6835

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

w13915984028 commented on pkg/webhook/resources/virtualmachine/validator.go:
```diff
@@ -646,3 +656,65 @@ func (v *vmValidator) checkEmptyMemory(vm *kubevirtv1.VirtualMachine) error {
+func (v *vmValidator) checkMaintenanceModeStrategyIsValid(newVM, oldVM *kubevirtv1.VirtualMachine) error {
+	newLabels := newVM.ObjectMeta.Labels
+	newStrategy, newOk := newLabels[util.LabelMaintainModeStrategy]
+	// Creating a new VM without maintenance mode strategy is not ok
```
In `rancher-manager` -> `Harvester` scenario, the created/updated VM, does not include this label

The `empty means default` might be acceptable

```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  annotations:
    harvesterhci.io/mac-address: '{"nic-0":"52:c1:db:cb:43:a2"}'
    harvesterhci.io/vmRunStrategy: RerunOnFailure
    harvesterhci.io/volumeClaimTemplates: >-
      [{"metadata":{"name":"gc5-pool1-sq6g8-q6z8l-disk-0-9tcd4","creationTimestamp":null,"annotations":{"harvesterhci.io/imageId":"default/image-q2rth"}},"spec":{"accessModes":["ReadWriteMany"],"resources":{"requests":{"storage":"20Gi"}},"storageClassName":"longhorn-image-q2rth","volumeMode":"Block"},"status":{}}]
    kubevirt.io/latest-observed-api-version: v1
    kubevirt.io/storage-observed-api-version: v1
  creationTimestamp: '2025-11-13T19:44:51Z'
  finalizers:
    - kubevirt.io/virtualMachineControllerFinalize
    - wrangler.cattle.io/VMController.CleanupPVCAndSnapshot
  generation: 1
  labels:
    guestcluster.harvesterhci.io/name: gc5
    harvesterhci.io/creator: docker-machine-driver-harvester
    harvesterhci.io/machineSetName: default-gc5-pool1
    harvesterhci.io/vmName: gc5-pool1-sq6g8-q6z8l
    nodepool.harvesterhci.io/name: pool1-sq6g8
  name: gc5-pool1-sq6g8-q6z8l
  namespace: default
```

m-ildefons replied:
Updated to accept creation of VMs which don't have the label at all.
Invalid values are still rejected, as well as updates which contain either invalid values or remove the label.

w13915984028 reviewed (commented):
The first question is still about the empty label, esp. on Rancher/Terraform or third party api scenario, forcely requires this label might break them.

w13915984028 reviewed (dismissed):
LGTM, thanks.

The tests cases are excellent.

m-ildefons committed:
Node Drain Helper: early return for non-VM Pods

Let the maintenance mode strategy filter return early for non-VM Pods.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

w13915984028 reviewed (approved):
LGTM, thanks.

martindekov reviewed (approved):
LGTM Moritz, my understanding is that now we allow only from the list of known and valid labels and reject bad formats and properly first taking pods part of the vm for migration and from those pods pick the strategy and skip if pod is not part of the vm.

Vicente-Cheng commented:
Thanks @m-ildefons for this PR.
If #9528 was too complex for v1.7.1, we can backport this to v1.7.1.

Vicente-Cheng commented:
> Why is this now marked DNM? Please explain.

The 1.7 was in code freeze stage.
Based on your description
> The scope of https://github.com/harvester/harvester/pull/9528 makes review too hard to complete in time for v1.7.0

I assume we should go back to review #9528 instead of this one.
Or you thought these changes were independent of #9528?
Please correct your statement.

m-ildefons commented:
This PR is independent of #9528

This PR contains a subset of changes from #9528. The changes in this PR had been identified by @w13915984028 as suitable for inclusion in v1.7.0, as discussed here: https://github.com/harvester/harvester/pull/9528#issuecomment-3572757277

Meanwhile #9528 contains further changes, which enforce the maintenance mode strategy behavior more aggressively. @w13915984028 raised concerns about being able to ensure correctness in time for the v1.7.0 release.

Therefore we decided that it would be best not to merge #9528 before the v1.7.0 release.
Instead I made this PR with only the changes that we are sure about, so we can at least have them for v1.7.0
All other changes can then receive the necessary work later.

w13915984028 commented:
I confirm that the current change ensures to fix the existing bug: when the expected label is wrong/missing, the maintain/drain controller might go in to wrong state. But allow the empty label to avod breking cases like docker-machine-driver which does not apply the maintain label at all on VM.

Please @votdev also double check the fix, thanks.

Vicente-Cheng reviewed (dismissed):
Just one nit, also nest if could be improved, but it did not affect the readability.
So it depends on you. 

LGTM thanks.

m-ildefons committed:
VirtualMachine Validator: Add early return

Return early in case mantenance mode strategy is missing during
create/update. This is possible because removing the maintenance mode
strategy will just cause the behavior to fall back to default.
If a VM manifest is sent to the API server (during creation or update of
a VM) and it doesn't have the maintenance mode strategy set, then we
don't need to do further checks at all.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

votdev reviewed (commented):
Even though the PR has already been merged; LGTM.
